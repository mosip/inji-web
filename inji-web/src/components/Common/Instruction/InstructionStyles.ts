export const InstructionStyles = {
    container:
        'bg-iw-paleGray rounded-xl flex flex-col items-start justify-center gap-2 p-4 pr-10 overflow-y-auto sm:max-h-[38vh] max-h-[35vh]',
    instructionContent: 'flex flex-col items-left gap-2 sm:gap-3 rounded-lg px-8 py-2',
    questionContainer: "flex gap-2 items-center align w-full",
    instructionQuestion:
        'text-xs sm:text-sm md:text-md text-left font-montserrat font-semibold text-black',
    instructionText:
        'text-xs sm:text-sm md:text-md text-left font-montserrat font-normal text-black list-disc pl-2 ',
    instructionTextStrong:
        'text-xs sm:text-sm md:text-md text-left font-montserrat text-black'
}