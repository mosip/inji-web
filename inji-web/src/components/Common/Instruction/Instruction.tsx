import React from "react";
import {InstructionStyles} from "./InstructionStyles";
import {InfoIcon} from "../Icons/InfoIcon";
import {InstructionItem} from "../../../types/data";

interface InstructionContentProps {
    instructionItems: InstructionItem[];
    testId: string;
}

const InstructionQuestion: React.FC<{ content: string, testId: string }> = ({content, testId}) => {
    return (
        <p className={InstructionStyles.instructionQuestion} data-testid={testId}>
            {content}
        </p>
    );
}

const InstructionContent: React.FC<InstructionContentProps> = ({
                                                                          instructionItems,
                                                                          testId
                                                                      }) => {
    return (
        <ul data-testid={testId} className={InstructionStyles.instructionContent}>
            {instructionItems.map((item) =>
                <li
                    key={item.id}
                    data-testid={item.id}
                    className={InstructionStyles.instructionText}
                >
                    {item.content}
                </li>
            )}
        </ul>
    );
};

export const Instruction: React.FC<{
    instructionItems: InstructionItem[];
    question: string;
    testId: string;
}> = ({instructionItems, question, testId}) =>
    <div className={InstructionStyles.container} data-testid={`text-${testId}-instruction`}>
        <div className={"flex gap-2 items-center align-middle w-full"}>
            <InfoIcon
                testId={`icon-${testId}-instruction`}
                className="mt-[0.1rem]"
            />
            <InstructionQuestion content={question} testId={`text-${testId}-question`}/>
        </div>
        <InstructionContent
            instructionItems={instructionItems}
            testId={`text-${testId}-instruction-content`}
        />
    </div>
