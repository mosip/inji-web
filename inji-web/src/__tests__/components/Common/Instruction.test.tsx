import React from 'react';
import { render, screen, within } from '@testing-library/react';
import { Instruction } from '../../../components/Common/Instruction/Instruction';

describe('Instruction', () => {
    const instructionItems = [
        { id: 'item-1', content: 'First instruction' },
        { id: 'item-2', content: <span>Second <strong>instruction</strong></span> }
    ];
    const question = 'What should I do?';
    const testId = 'test-instruction';

    it('should match instruction component snapshot', () => {
        const { asFragment } = render(
            <Instruction
                instructionItems={instructionItems}
                question={question}
                testId={testId}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders the question with correct text and test id', () => {
        render(
            <Instruction
                instructionItems={instructionItems}
                question={question}
                testId={testId}
            />
        );

        expect(screen.getByText(question)).toBeInTheDocument();
        expect(screen.getByTestId(`text-test-instruction-question`)).toBeInTheDocument();
    });

    it('renders the info icon with correct test id', () => {
        render(
            <Instruction
                instructionItems={instructionItems}
                question={question}
                testId={testId}
            />
        );

        expect(screen.getByTestId(`icon-test-instruction-instruction`)).toBeInTheDocument();
    });

    it('renders the instruction list with correct role and test id', () => {
        render(
            <Instruction
                instructionItems={instructionItems}
                question={question}
                testId={testId}
            />
        );

        expect(screen.getByRole('list')).toBeInTheDocument();
        expect(screen.getByTestId(`text-test-instruction-instruction-content`)).toBeInTheDocument();
    });

    it('renders all instruction items with correct text and test ids', () => {
        render(
            <Instruction
                instructionItems={instructionItems}
                question={question}
                testId={testId}
            />
        );

        expect(screen.getByTestId("item-1")).toBeInTheDocument();
        expect(screen.getByText('First instruction')).toBeInTheDocument();
        expect(screen.getByTestId("item-2")).toBeInTheDocument();
        const secondInstruction = screen.getByTestId('item-2');
        expect(within(secondInstruction).getByText('Second')).toBeInTheDocument();
        expect(within(secondInstruction).getByText('instruction', { selector: 'strong' })).toBeInTheDocument();
    });

    it('renders the container with correct test id', () => {
        render(
            <Instruction
                instructionItems={instructionItems}
                question={question}
                testId={testId}
            />
        );

        expect(screen.getByTestId(`text-test-instruction-instruction`)).toBeInTheDocument();
    });
});