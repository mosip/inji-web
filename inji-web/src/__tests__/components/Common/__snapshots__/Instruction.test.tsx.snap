// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Instruction should match instruction component snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-iw-paleGray rounded-xl flex flex-col items-start justify-center gap-2 p-4 pr-10 overflow-y-auto sm:max-h-[38vh] max-h-[35vh]"
    data-testid="text-test-instruction-instruction"
  >
    <div
      class="flex gap-2 items-center align-middle w-full"
    >
      <img
        alt="Info"
        class="mt-[0.1rem]"
        data-testid="icon-test-instruction-instruction"
        src="InfoIcon.svg"
      />
      <p
        class="text-xs sm:text-sm md:text-md text-left font-montserrat font-semibold text-black"
        data-testid="text-test-instruction-question"
      >
        What should I do?
      </p>
    </div>
    <ul
      class="flex flex-col items-left gap-2 sm:gap-3 rounded-lg px-8 py-2"
      data-testid="text-test-instruction-instruction-content"
    >
      <li
        class="text-xs sm:text-sm md:text-md text-left font-montserrat font-normal text-black list-disc pl-2 "
        data-testid="item-1"
      >
        First instruction
      </li>
      <li
        class="text-xs sm:text-sm md:text-md text-left font-montserrat font-normal text-black list-disc pl-2 "
        data-testid="item-2"
      >
        <span>
          Second 
          <strong>
            instruction
          </strong>
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
